README.md
# crawler_toolkit: webdriver action & output_data function
###### tags: `Python`, `pypi`

## What is it?
**crawler_toolkit** is a python package that provides some use function for control webdriver, and easy outputdata to mongodb, mssql, solr, json file.


## Dependencies
- [pandas](https://pypi.org/project/pandas/)
- [tqdm](https://pypi.org/project/tqdm/)
- [Sqlalchemy](https://pypi.org/project/SQLAlchemy/)
- [Sqlalchemy-utils](https://pypi.org/project/SQLAlchemy-Utils/)
- [pymongo](https://pypi.org/project/pymongo/)
- [pydbc](https://pypi.org/project/pyodbc/)
- [pysolr](https://pypi.org/project/pysolr/)
- [selenium-wire](https://pypi.org/project/selenium-wire/)
- [opencc-python-reimplemented](https://pypi.org/project/opencc-python-reimplemented/)
- [scrapy](https://pypi.org/project/Scrapy/)
- [stem](https://pypi.org/project/stem/)
- [undetect-chromedriver](https://pypi.org/project/undetected-chromedriver/)

## Other driver need
- [ODBC Driver](https://learn.microsoft.com/zh-tw/sql/connect/odbc/download-odbc-driver-for-sql-server?view=sql-server-ver16)

ChangeLog.md
# Change Log: crawker_toolkit
###### tags: `ChangeLog`, `python-package`

### `2023-05-02` **v0.0.8**: 
- **mailserver.GmailServer**
    - add:
        - `set_images()`
- **share_module.Message**
    - update:
        - `sleeptime()`: only accept int or float
- update Descriptions:
    - generator.py 
    - mailserver.py
    - share_module.py
    - tor_controller.py

### `2023-04-26` **v0.0.7a0** hotfix: output_data.mongodb_actions
- **`MongoDB_Actions`**
    - fixed: 
        - remove print(error) and write to log
        - mongodb_action.retrieve add args how to select find_all or find_one
- **`mailserver`**
     - add:
         - mailserver add function description
         - `GMailServer.mail_prepare`: accept str or list mails will prepare object that can send by mailserver
- **`TempDataSet`**
    - refactor
        - replace name from `DuplicateFilter` to `TempDataSet`
        - `self.is_duplicate`: rename passing argument `url`> `unique_val`


### `2023-04-18` **v0.0.7** feature: **refactor output_data**
- **`MssqlActions`**
    - add:
        - __repr__, __str__
        - functions: `syntax_query`, `drop_table`, `check_connection`
        - refactor: create for single and multiple data insert
        - all function description
        - unittest for mssql_actions : tests/mssql_action.py
    - fixed:
        - rename func retreive > retrieve
        - create_all merge with create
        - update remove update UpdatedTime
        - delete remove update DeteltedTime
        - rename _create_connection_url > create_connection_url
        - create_connection_url to classmethod > staticmethod
- **`SolrActions`**
    - add:
        - __repr__, __str__
        - function: `check_connection`
        - unittest for solr_actions : tests/solr_actions.py
    - update:
        - all function description
- **`FileActions`**
    - add:
        - unittest for file_acitons : tests/file-actions.py
        - all function description
    - fixed:
        - toPickle & LoadPicke not close
- **`MongoActions`**
    - add:
        - __repr__, __str__, check_connection, create, update, delelte, drop_colleciton
        - unittest for solr_actions : tests/mongodb_actions.py
    - update:
        - all function description
        - rename `retreive` > `retrieve`
        - rename `create_update_exist` > `upsert`

### `2023-01-31` **v0.0.6**
- add:
    - **webdriver_action.init_driver** args:
        - `use_driver`: str=add "remote" options
        - `remote_driver`: str=passing remote driver use name: "firefox", "chrome", "edge"
        - `set_preference`: dict=passing dictionary preference k,value, for this time passing tor socket information setup proxy
        - `seleniumwire_option` replace "proxy_dict": dict= proxy_dict passing into seleniumwire_option, and seleniumwire option more available args can use.
    - **tor_controller**:
        - `TorController.switch_ip`: passing tor ip:port and auth_password to control tor to change ip
    
### `2023-01-06` **v0.0.5**
- add:
    - **generator**
        - `FakeIdentity.user`: generator user `name_zh`, `name_en`, `age`, `birthdate`, `account_name`, `address_zh`, `address_en`
        - `FakeIdentity.password`: generator password with random charactor in `lower, upper, numbers`
        - `UserAgent.get_latest`: scrapy latest user agent in website
        - `UserAgent.browser_list`: show available browser.
    - **webdriver_action**
        - `WebdriverAction.send_keys`: webdriver input value into element
        - `WebdriverActions.get_element_properties`: return all property of tag element 
- update:
    - **output_data**
        - `Mssql_Actions.__row2dict` > `Mssql_Actions.row2dict`: open to use
- remove:
    - **webdriver_action**
        - `WebdriverActions.find_elem_xpath`: new version use `WebdriverActions.find_elem_by`
        

### `2022-12-27` **v0.0.4**
- add:
    - **output_data**
        - `Files_Actions.readPickle`: load pickle files from path
        - `Files_Actions.toPickle`: saving pickle files to path
- update:
    - **webdriver_action**
        - `WebdriverAction.loading_cookies`: split `pickle` loading function

### `2022-12-22` **v0.0.3~a2**
- add:
    - **mailserver**:
        - `GMailServer`: Send Mail By Gmailserver
    - **webdriver_action**: 
        - `WebdriverAction.load_cookies`: load pickle_files to webdriver
        - `WebdriverAction.update_cookies`: webdrvier update cookie to pickle_files
        - `WebdriverAction.fetch_auth_key`: fetch two-factor key
        - `WebdriverAction.find_element_by`: webdriver find element by multiple select option args.
        - `ResponseProcess.decode_body`: decode webdriver response
    - **share_module**:
        - `DataProcess.str2hash`
        - `ArgsTF_define`: `TF_define`, `cama_str2list`

### `2022-12-15` **v0.0.2.a**
- hotfix:
    - `output_data.MsSql_Action.create_all()`: IntegrityError loop stop
    - `webdriver_action.WebdriverAction.init_driver()`: headess and option not work
            
### `2022-11-14` **v0.0.2**
- add:
    - **share_module**:
        1. `DataProcess` dictionary nest_get function and chinese translate etc... funtion
        2. `ArgsDefine` to deal with cmd passing args
        3. `DuplicationFilter` put string value into tuple that check already use value return `T/F`
        4. `File_Action` add `readJson`, `updateJson`
        5. `File_Action.toJson` add args append to control data append or over write all files
- update:
    - **webdriver_action**:
        1. `WebdriverAction.init_driver` to init driver

### `2022-11-10` **v0.0.1**
init project
- add:
    1. **share_module**:
        - `LoggingSet` that setting logging for project
        - `Message` that provide to print an message with custom splitline & sleep time with message for webdriver waiting.
    2. **output_data**:
        - `File_Action` to output data to file
        - `MongoDB_Action` to output data to mongodb
        - `MsSQL_Actions` to output data to Sql server

tests/file_actions.py
import json
import unittest

import os
import sys
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from src.crawler_toolkit.output_data import Files_Actions

class FileActionTest(unittest.TestCase):
    filter_by = {
        'id': 99
    }

    single_data = {
        'id': 99,
        'name': 'test_solger'
    }

    multiple_data = [ {'id': i, 'name': f'test_{i}'} for i in range(4)]
    test_filefolder = './tests/fileaction_test'
    if not os.path.exists(test_filefolder):
        os.mkdir(test_filefolder)
    file_name = os.path.join(test_filefolder, 'test.json')
    pickle_name = os.path.join(test_filefolder, 'pickle.pkl')
    
    def test_1_check_filterkey(self):
        check_bool = Files_Actions.check_filter_key(
                        dict_data=self.single_data,
                        filter_by=self.filter_by
                    )
        self.assertTrue(check_bool, 'Fail check Filter key')
    
    def test_2_tojson_single(self):
        Files_Actions.toJson(output_data=self.single_data, output_path=self.file_name)
        result = Files_Actions.readJson(file_path=self.file_name)
        self.assertEqual(result, [self.single_data])

    def test_3_tojson_multiple(self):
        Files_Actions.toJson(output_data=self.multiple_data, output_path=self.file_name)
        result = Files_Actions.readJson(file_path=self.file_name)
        self.assertGreater(len(result), 4)

    def test_4_readjson(self):
        result = Files_Actions.readJson(file_path=self.file_name)
        self.assertGreater(len(result), 0)

    def test_5_updatejson_single(self):
        update_data = [{'filter_by': {'id': 99}, 'update_data': {'name': 'updateName'}}]
        Files_Actions.updateJson(update_data=update_data, file_path= self.file_name)
        result = Files_Actions.readJson(file_path=self.file_name)
        for r in result:
            if r.get('id') == 99:
                self.assertEqual(r.get('name'), 'updateName', 'Update Fail')
    
    def test_5_updatejson_multiple(self):
        mul_update_data = [
            {'filter_by': {'id': 99}, 'update_data': {'name': 'updateName_1'}},
            {'filter_by': {'id': 0}, 'update_data': {'name': 'updateName_1'}}
            ]
        Files_Actions.updateJson(update_data=mul_update_data, file_path= self.file_name)
        result = Files_Actions.readJson(file_path=self.file_name)
        for r in result:
            if r.get('id') in [99, 0]:
                self.assertEqual(r.get('name'), 'updateName_1', 'Update Fail')

    def test_6_loadPickle(self):
        Files_Actions.toPickle(output_data={'cookies': 'asflkasjdinb', 'header': 'google_chrome 111'}, filepath=self.pickle_name)
        self.assertTrue(os.path.exists(self.pickle_name), 'init pickle Fail')

    def test_7_toPickle(self):
        result = Files_Actions.loadPickle(filepath=self.pickle_name)
        self.assertEqual(result.get('cookies'), 'asflkasjdinb')

    def test_8_remove_folder(self):
        os.remove(self.file_name)
        os.remove(self.pickle_name)
        os.rmdir(self.test_filefolder)

if __name__ == '__main__':
    unittest.main()

tests/mssql_action.py
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from src.crawler_toolkit.output_data import MsSQL_Actions

import unittest
from datetime import datetime

from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from sqlalchemy import Column, ForeignKey
from sqlalchemy.dialects.mssql import INTEGER, BIT, DATETIME, VARCHAR, TEXT, NVARCHAR
from sqlalchemy.ext.declarative import declarative_base

db = declarative_base()

class TestTable(db):
    """ TestTable """
    __tablename__ = "TestTable"
    Id   = Column(INTEGER, primary_key=True, autoincrement=True)
    AccountType = Column(VARCHAR(32), nullable=False)
    Created_Time = Column(DATETIME, default=func.now())


class MssqlActionTest(unittest.TestCase):
    TEST_CONNECTION = {
            'username': None,
            'password': None,
            'host': None,
            'port': None,
            'database': 'develop/sqlite.db',
            'drivername': 'sqlite',
            'query': None
        }
    ROOT_PATH = os.path.dirname(os.path.dirname(__file__))
    sql_action = MsSQL_Actions(connect_info=TEST_CONNECTION)
    create_time = datetime.strptime('2023-04-12 15:00:00','%Y-%m-%d %H:%M:%S')
    
    def test_1_connection(self):
        self.assertTrue(self.sql_action.check_connection(), 'Connection Fail')

    def test_2_init_database_table(self):
        self.sql_action.init_database(TestTable)
        db_path = os.path.join(self.ROOT_PATH, self.TEST_CONNECTION.get('database'))
        self.assertTrue(os.path.exists(db_path), 'Database Init Fail, DB not found')

    def test_3_table_exist(self):
        table_schema = self.sql_action.syntax_query('SELECT name FROM sqlite_schema')
        self.assertEqual(table_schema, [{'name': 'TestTable'}], 'Table Schema Not Exist')

    def test_4_create_single_data(self):
        insert_data = {'AccountType': 'Test1', 'Created_Time': self.create_time}
        self.sql_action.create(datas=insert_data, table_model=TestTable)
        sqry = 'SELECT * FROM TestTable WHERE AccountType = "Test1"'
        result = self.sql_action.syntax_query(sqry)
        self.assertGreater(len(result), 0, 'create_data fail')

    def test_5_create_multiple_data(self):
        # init fake data
        insert_data = list()
        for i in range(10):
            insert_data.append({'AccountType': f'type_{i}', 'Created_Time': self.create_time})
        self.sql_action.create(datas=insert_data, table_model=TestTable)

        # get account_type values
        type_list = [i['AccountType'] for i in insert_data]
        
        # retrieve from db to check create data len equal
        query_dict = {'AccountType': type_list}
        result = self.sql_action.retrieve(table_model=TestTable, filter_by_dict=query_dict)
        self.assertEqual(len(result), len(insert_data), 'create_data multiple fail')

    def test_6_update(self):
        self.sql_action.update(table_model=TestTable, data_id=1, update_data={'AccountType': 'Test_Modify'})
        result = self.sql_action.retrieve(table_model=TestTable, filter_by_dict={'Id': 1})
        if result:
            self.assertEqual(result[0]['AccountType'], 'Test_Modify', 'update value Fail')
        else:
            raise ValueError('empty result')
        
    def test_7_delete(self):
        self.sql_action.delete(table_model=TestTable, data_id=1)

        result = self.sql_action.retrieve(table_model=TestTable, filter_by_dict={'Id': 1})
        self.assertEqual(result, [], 'delete record Fail')

    def test_8_drop_table(self):
        # drop table
        self.sql_action.drop_table(table_model=TestTable)

        # select table exist or not
        table_schema = self.sql_action.syntax_query('SELECT name FROM sqlite_schema WHERE name = "TestTable"')
        self.assertEqual(table_schema, [], 'Table Schema Exist, Drop Fail')

if __name__ == '__main__':
    unittest.main()

tests/mongodb_actions.py
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from src.crawler_toolkit.output_data import MongoDB_Actions


import unittest

class MongoDBActionsTest(unittest.TestCase):
    MONGODB_CONN_INFO = {
        'database_name': 'TEST',
        'host': '127.0.0.1',
        'port': 27017,
        'username': 'admin',
        'password': 'asdf',
        'auth_source': 'admin'
    }

    mongo_actions = MongoDB_Actions(connect_info=MONGODB_CONN_INFO)
    fake_data = [{'_id': i,'name': f'test_{i}', 'age': i*10} for i in range(4)]
    collection_name = 'unittest'
    def test_1_check_connection(self):
        self.assertTrue(self.mongo_actions.check_connection(), 'Connection Fail')

    def test_2_create(self):
        self.mongo_actions.create(create_data=self.fake_data, collection_name=self.collection_name)
        result = self.mongo_actions.retrieve(collection_name=self.collection_name, query_syntax={})
        self.assertEqual(result, self.fake_data, 'Create Data Fail')
    
    def test_3_update(self):
        query_syntax = {'_id': 0}
        update_result = self.mongo_actions.update(update_data={'name': 'test_update'}, query_syntax=query_syntax, collection_name=self.collection_name)
        result = self.mongo_actions.retrieve(collection_name=self.collection_name, query_syntax=query_syntax)[0]
        self.assertEqual(result.get('name'), 'test_update', 'Update Data Fail')

    def test_4_upsert(self):
        upsert_data = [{'_id': 3, 'name': 'upsert_test', 'age': 30}, {'_id': 4, 'name': 'upsert_test', 'age': 40}]
        self.mongo_actions.upsert(create_data=upsert_data, collection_name=self.collection_name)
        result = self.mongo_actions.retrieve(collection_name=self.collection_name, query_syntax={'name': 'upsert_test'})
        self.assertEqual(result, upsert_data, 'Upsert=true, Fail')

    def test_5_delete(self):
        query_syntax = {'_id': 0}
        self.mongo_actions.delete(collection_name=self.collection_name, query_syntax=query_syntax)
        result = self.mongo_actions.retrieve(collection_name=self.collection_name, query_syntax=query_syntax)
        self.assertEqual(len(result), 0, 'Delete data Fail')

    def test_9_drop_collection(self):
        drop_success = self.mongo_actions.drop_collection(collection_name=self.collection_name)
        self.assertTrue(drop_success, 'Drop collection Fail')

        drop_success = self.mongo_actions.drop_collection(collection_name=self.collection_name)
        self.assertFalse(drop_success, 'collection not exist')

if __name__ == '__main__':
    unittest.main()

src/crawler_toolkit/mailserver.py
import re
import smtplib
from os.path import basename
from crawler_toolkit.share_module import Message
from email.mime.image import MIMEImage
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from email.mime.application import MIMEApplication
from email.utils import COMMASPACE, formatdate


class GMailServer:
    '''Gmail Server

    It is an Mail object which you can structure your mail 
    and send to your target by SMTP mailserver

    :param - account <str>: host to send mail account 
    :param - api_key <str>: SMTP access token api key
    :param - send_to <list|str>: mail receiver
    :param - cc <list|str>: carbon copy receiver
    '''
    def __init__(self, 
            account: str, 
            api_key: str, 
            send_to: list, 
            cc: list=None, 
        ):
        self.account = account
        self.api_key = api_key
        self.send_to = self.mail_prepare(send_to)
        self.cc = self.mail_prepare(cc)
        self.msg = MIMEMultipart()

    @staticmethod
    def mail_prepare(mails :str|list)-> str:
        '''mail_prepare 

        accept str or list mails will prepare object that can send by mailserver
        
        :param - mails <str| list>: passing mail address.
        '''
        if mails:
            if not isinstance(mails, (str, list)):
                raise TypeError('mails only accept type [str| list]')
            if isinstance(mails, str):
                if re.search(',', mails):
                    mails = mails.split(',')
                else:
                    mails = [mails]
        return mails

    def set_mail_info(self, subject: str):
        self.msg['From'] = self.account
        self.msg['To'] = COMMASPACE.join(self.send_to)
        if self.cc:
            self.msg['Cc'] = COMMASPACE.join(self.cc)
        self.msg['Date'] = formatdate(localtime=True)
        self.msg['Subject'] = subject


    def set_html(self, html):
        ''' set html content to mail

        :param - html <str>: html string content to add into mail object
        '''
        self.msg.attach(MIMEText(html, "html"))
    
    def set_text(self, text):
        ''' set text content to mail

        :param - text <str>: string content to add into mail object
        '''
        self.msg.attach(MIMEText(text))

    def set_image(self, filelist_path: list[str]):
        ''' set file img to mail

        :param - filelist_path <list[str]>: set images files to mail from list filepath want to send by mail
        '''
        for filepath in filelist_path:
            with open(filepath, "rb") as f:
                part = MIMEImage(
                    f.read(),
                    Name=basename(filepath)
                )
            # After the file is closed
            part['Content-Disposition'] = 'attachment; filename="%s"' % basename(filepath)
            self.msg.attach(part)

    def set_files(self, filelist_path: list[str]):
        ''' set file obj to mail

        :param - filelist_path <list[str]>: set file to mail from list filepath want to send by mail
        '''
        for filepath in filelist_path:
            with open(filepath, "rb") as f:
                part = MIMEApplication(
                    f.read(),
                    Name=basename(filepath)
                )
            # After the file is closed
            part['Content-Disposition'] = 'attachment; filename="%s"' % basename(filepath)
            self.msg.attach(part)
    
    def send_mail(self, host: str='imap.gmail.com', port: int=465):
        ''' send mail to send_to and cc target mail
        
        :param - host <str>: mail server SMTP_SSL host
        :param - port <int>: mail server SMTP_SSL port
        '''
        Message.splitline('[MailServer]Send Mail', f'from {self.account} To: {self.send_to} CC: {self.cc}')
        with smtplib.SMTP_SSL(host, port) as smtp_server:
            total_send = self.send_to
            if self.cc:
                total_send = total_send + self.cc
            smtp_server.login(self.account, self.api_key)
            smtp_server.sendmail(self.account, total_send, self.msg.as_string())

src/crawler_toolkit/share_module.py
import re
import os
import sys
import logging
from tqdm import tqdm
from time import sleep
from opencc import OpenCC
from hashlib import sha256
from datetime import datetime


class LoggingSet:
    def setup_applevel_logger(logger_name: str='logging', file_name=None, log_level: str='debug', log_path: str=None):
        '''setup applevel logger
        init log and settings error level path etc..

        :param - `logger_name` <str>: loggername
        :param - `filename` <str>: output log filename
        :param - `log_level` <str>: setting log level [info| debug| warning| error]
        :param - `log_path`: <str>: settings output log folder path
        '''
        log_path = log_path or os.path.dirname(os.path.abspath(__name__))
        log_dict = {
                'info'   : logging.INFO,
                'debug'  : logging.DEBUG,
                'warning': logging.WARNING,
                'error'  : logging.ERROR
            }
        logger = logging.getLogger(logger_name)
        logger.setLevel(log_dict[log_level])
        formatter = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
        sh = logging.StreamHandler(sys.stdout)
        sh.setFormatter(formatter)
        logger.handlers.clear()
        logger.addHandler(sh)
        if file_name:
            fh = logging.FileHandler(os.path.join(log_path, file_name))
            fh.setFormatter(formatter)
            logger.addHandler(fh)
        return logger

    def get_logger(module_name: object, logger_name: str='logging'):
        '''
        set logger object for module

        :param- `module_name` <module object>: passing class or function object to init child logger to output log
        :param- `logger_name` <str>: passing same name as self.setup_applevel_logger settings "logger_name" args
        '''
        return logging.getLogger(logger_name).getChild(module_name)


class Message:
    @staticmethod
    def splitline(title: str , context: str, linesymbol: str = '#')-> str:
        '''splitline 
        init split line with passing line symbol and print message
        
        :param - `title` <str>: splitline title
        :param - `context` <str>: infomation which wanna print out
        :param - `linesymbol` <str>: splitline use passing symbol, default is #

        preview example: \n
            |####| title |####|\n
            context
        '''
        print()
        print(str(datetime.now()) + ('|{:' + linesymbol + '^75}|').format('| %s |'%title))
        print(context)

    @staticmethod
    def sleep_time(seconds: int, custom_message: str='sleep action start'):
        '''sleep_time
        splitline and wait for seconds.

        :param - `seconds` <int>: wait for seconds
        :param - `custom_message` <str>: settings print information, default= "sleep action start"
        '''
        Message.splitline(f'[Message] total sleep: {seconds}', custom_message)
        if not isinstance(seconds, (int, float)):
            raise TypeError('args "seconds" only accept type <int>')
        if isinstance(seconds, float):
            seconds = int(seconds)
        
        time_loop = tqdm(range(seconds))
        for i in time_loop:
            sleep(1)


class DataProcess:
    @staticmethod
    def word_translate(str_data: str, how: str='t2s')-> str:
        '''word_translate
        it is transfer chinese between traditional and simpleified 
        
        
        :param - `str_data` <str>: which data need to transfer
        :param - `how` <str>: from which chinese to others, default="t2s"
            :`how` more option in `openCC` pypi url: https://pypi.org/project/opencc-python-reimplemented/
        '''
        cc = OpenCC(how)
        if str_data :
            str_data  = cc.convert(str_data)
        return str_data

    @staticmethod
    def nested_get(input_dict: dict, nested_key: list):
        '''nested_get
        extract value with nested_key from input_dict

        :param - `input_dict` <dict>: nested dictionary data
        :param - `nested_key` <list>: list level key
        '''
        result = input_dict
        if isinstance(input_dict, list):
            return None
        for k in nested_key:
            result = result.get(k, None)
            if result is None:
                return None
        return result

    @staticmethod
    def str2hash256(data_byte: bytes):
        '''string data to hash256

        :param `data_byte` <bytes>: data which want to trans to hash256
        '''
        if isinstance(data_byte, str):
            data_byte = data_byte.encode()
        hash_str = sha256()
        hash_str.update(data_byte)
        return hash_str.hexdigest()


class TempDataSet:
    '''temp data set
    
    init data_set to store unique value
    '''
    def __init__(self) -> None:
        self.data_set = set()
    
    def is_duplicated(self, unique_val: str)-> bool:
        '''check data already in set return True
        False if not in set and append unique_val to set

        :param - `unique_val` <str>: duplicate detect by `unique_val`
        '''
        if unique_val in self.data_set:
            return True
        self.data_set.add(unique_val)
        return False


class ArgsDefine:
    def TF_define(passing_args: str)-> bool:
        '''define string true false
        
        :param - `passing_args`: str: string true false or 0 to bool
        '''
        if passing_args in ['True', 'true', '1']:
            return True
        elif passing_args in ['False', 'false', '0']:
            return False
        else:
            raise ValueError(' only accept [True|False]')

    def cama_str2list(args_str: str):
        '''split string with cama to list data

        :param - `args_str` <str>: string data which want to split
        '''
        args_str = re.sub(' ', '', args_str)
        return args_str.split(',')


class DefaultSet:
    def init_folder(folder_list: list[str])-> None:
        '''init filepath 
        
        :param - folder_list <list[str]>: list default path need to create
        '''        
        for folder in folder_list:
            if folder and not os.path.exists(folder):
                os.mkdir(folder)
                print(f'path create: {folder}')


src/crawler_toolkit/generator.py
import os
import re
import json
import string
import random
import requests
import pandas as pd
from scrapy import Selector
from datetime import date, datetime, timedelta

class FakeIdentity:
    ROOT_PATH = os.path.dirname(__file__)
    ZH_DATA_SOURCE = os.path.join(ROOT_PATH, 'source', 'zh_data')
    EN_DATA_SOURCE = os.path.join(ROOT_PATH, 'source', 'en_data')

    def en_name(gender: str):
        '''random init english name by gender
        
        :param - `gender` <str>: passing gender to init name, ["male", "female"]
        '''
        with open(os.path.join(FakeIdentity.EN_DATA_SOURCE, 'en_name.json'), 'r') as f:
            en_data_source = json.load(f)
        last_name = random.choice(en_data_source['lastname']).lower()
        first_name = random.choice(en_data_source[gender]).lower()
        return {
            'name': ', '.join([first_name , last_name]),
            'last_name' : last_name,
            'first_name' : first_name,
            'gender': gender
        }

    def cn_name()-> dict:
        '''random init chinese name
        '''
        result = dict()
        lastname_file = pd.read_json(os.path.join(FakeIdentity.ZH_DATA_SOURCE, 'lastname.json'))
        lastname_file = lastname_file.rename(columns={i:'LastName_'+i for i in lastname_file.columns})
        givename_file = pd.read_json(os.path.join(FakeIdentity.ZH_DATA_SOURCE, 'givename.json'))
        givename_file = givename_file.rename(columns={i:'GiveName_'+i for i in givename_file.columns})
        lastname = random.choice(lastname_file.to_dict('records'))
        givename = random.choice(givename_file.to_dict('records'))
        result.update(lastname)
        result.update(givename)
        return result

    def address()-> dict:
        '''random init taiwan address in zh & en
        '''
        city_info = pd.read_json(os.path.join(FakeIdentity.ZH_DATA_SOURCE, 'city.json'))
        street_name = pd.read_json(os.path.join(FakeIdentity.ZH_DATA_SOURCE, 'road.json'))
        street = {'zh_name': None, 'en_name': None}
        city = random.choice(city_info.to_dict('records'))
        road_of_city = street_name[street_name['site_id'] == (city['zh_city']+ city['zh_dist'])]
        if not road_of_city.empty:
            street = random.choice(road_of_city.to_dict('records'))
        nums = random.choice(range(1, 100))
        
        zh_address = ', '.join([city.get('zh_city',''), f"{city.get('zh_dist', '')} {street.get('zh_name','')} {nums}號"])
        en_address = ', '.join([f'No. {nums}', f"{street.get('en_name','')}, {city.get('en_dist','')}, {city.get('en_city','')}"])
        return {
            'address_zh' : re.sub('None', '', zh_address),
            'address_en' : re.sub('None', '', en_address),
        }

    def experience(age: int, birthdate: str, address: str=None)-> dict:
        '''random college & company experience
        '''
        result = {
            'CollegeName': None,
            'Department' : None,
            'CollegeTimeRange': None,
            'CompanyName' : None,
            'CompanyAddress': None,
            'WorkTimeRange': None,
        }
        cmpy = pd.DataFrame()
        college = pd.read_json(os.path.join(FakeIdentity.ZH_DATA_SOURCE, 'tw_college.json'))
        cmpy_json = pd.read_json(os.path.join(FakeIdentity.ZH_DATA_SOURCE, 'tw_company.json'))

        college =random.choice(college.to_dict('records')) if age >=18 else None
        if college:
            age18 = datetime.strptime(birthdate, '%Y-%m-%d') + timedelta(days=365*18)
            school_startyear = age18.year+1 if age18.month>9 else age18.year
            school_endyear = f'~{school_startyear+4}-06' if date.today().year - school_startyear >=4 else ''
            result.update({
                'CollegeName' : college.get('name_zh'),
                'Department'  : college.get('department_name'),
                'CollegeTimeRange': f'{school_startyear}-09{school_endyear}',
            })

            if address:
                cmpy = cmpy_json[cmpy_json['address'].str.match(f'{address}')]
            if cmpy.empty:
                cmpy = cmpy_json
            cmpy = random.choice(cmpy.to_dict('records')) if age > 22 else None

            if cmpy and school_endyear != '':
                cmp_startyear = school_startyear+5
                result.update({
                    'CompanyName' : cmpy.get('company_name'),
                    'CompanyAddress': cmpy.get('address'),
                    'WorkTimeRange': str(date.today().replace(year=cmp_startyear)),
                })

        return result

    def user(start_age: int=15, end_age: int=45):
        '''init user from range(`start_age` to `end_age`)
        return user with name age gender birthdate account name and address
        '''

        name_object = FakeIdentity.cn_name()
        age = random.choice(range(start_age, end_age))
        dateborn = random.choice(range(365))
        birthdate = date.today() - timedelta(days=age*365) - timedelta(days=dateborn)
        birthdate = str(birthdate)
        random_nums= birthdate.split('-')[-1] + str(random.choice(range(99)))

        zh_name = name_object.get(f'LastName_Character') + name_object.get(f'GiveName_Character')
        en_name = ','.join([name_object.get(f'LastName_English'),name_object.get(f'GiveName_English')])
        account_name = re.sub(',|-','',en_name) + random_nums

        address = FakeIdentity.address()
        adderss_dist = address.get('address_zh').split(',')[0]
        experience = FakeIdentity.experience(age=age, birthdate=birthdate, address=adderss_dist) 

        user = {
                'Name_zh': zh_name,
                'Name_en': en_name,
                'Age': age,
                'Gender'  : name_object.get(f'GiveName_Gender'),
                'BirthDate': birthdate,
                'AccountName' : account_name.lower(),
                'Address_zh': address.get('address_zh'),
                'Address_en': address.get('address_en')
            }

        if experience:
            user.update(experience)
        return user

    def account(nickname: str,  with_nums=True)-> str:
        '''random init account by nickname
        
        :param - nickname <str>: passing english nick name to init account
        :param - with_nums <bool>: account with number, default=True
        '''
        ran_nums_char = random.choice(range(4)) or 2
        ran_char = ''.join(random.choices(string.ascii_lowercase, k=ran_nums_char))
        create_list = [nickname, ran_char]
        if with_nums:
            ran_nums = ''.join(random.choices(string.digits, k=ran_nums_char))
            create_list = [nickname, ran_char, ran_nums]
        return ''.join(create_list)

    def password()-> str:
        '''random init password
        '''
        password_length = len(string.ascii_lowercase)
        half_length = password_length //3
        lower = string.ascii_lowercase
        upper = string.ascii_uppercase
        num = string.digits
        charater = ''.join([lower , upper])
        charater = random.sample(lower , half_length + password_length % 3)
        charater += random.sample(upper , half_length )
        charater += random.sample(num , half_length)
        temp = random.sample(charater , password_length)
        return ''.join(temp)


class UserAgent:
    def get_latest(browser: str)-> dict:
        '''scrape select browser  list dictionary useragent object 
        
        :param - `browser` <str>: passing browser which want to scrape
        '''
        total_data = pd.DataFrame()
        target_url = f'https://www.whatismybrowser.com/guides/the-latest-user-agent/{browser}'

        r = requests.get(target_url)
        page_source = Selector(text=r.text)
        total_tag = page_source.xpath('.//section[contains(@class, "section-block-main-extra")]/div/div[@class="content-block-main"]/*')
        for idx, tag in enumerate(total_tag):
            if tag.xpath('self::table[contains(@class, "listing-of-useragents")]'):
                title = total_tag[idx-1].xpath('text()').get()
                table_df = pd.read_html(total_tag[idx].get())[0]
                table_df.columns = ['Platform', 'User_Agent']
                table_df['Browser'] = browser
                table_df['System'] = title

                total_data = pd.concat([total_data, table_df])
        return total_data.to_dict(orient='records')

    def browser_list()-> list[str]:
        '''return list browser scrape options
        
        '''
        return ['chrome', 'firefox', 'safari', 'internet-explorer', 'edge', 'opera', 'vivaldi', 'yandex-browser']